tests/components/test_nlp.py

```diff
@@ -316,6 +316,67 @@ def test_allowed_entities(kwik_e_mart_nlp):
     assert res['entities'][0]['text'] == 'xyz'
 
 
+def test_disallowed_entities(kwik_e_mart_nlp):
+    res = kwik_e_mart_nlp.process("hello")
+    assert res['intent'] == 'greet'
+    res = kwik_e_mart_nlp.process("hello", deny_nlp=["store_info.greet"])
+    assert res['intent'] != 'greet'
+
+    res = kwik_e_mart_nlp.process("transfer $200 from checking to savings")
+    assert res['intent'] == 'transfer_money'
+    assert res['entities'][0]['type'] == 'sys_amount-of-money'
+    assert res['entities'][1]['type'] == 'account_type'
+    assert res['entities'][1]['role'] == 'origin'
+    assert res['entities'][2]['type'] == 'account_type'
+    assert res['entities'][2]['role'] == 'dest'
+
+    res = kwik_e_mart_nlp.process("transfer $200 from checking to savings",
+                                  deny_nlp=["banking.transfer_money"])
+    assert res['intent'] != 'transfer_money'
+
+    res = kwik_e_mart_nlp.process("transfer $200 from checking to savings",
+                                  deny_nlp=["banking.transfer_money.account_type"])
+    assert res['intent'] == 'transfer_money'
+    assert 'account_type' not in {entity['type'] for entity in res['entities']}
+
+    res = kwik_e_mart_nlp.process("please can you tell me if springfield is "
+                                  "possibly open at this time on friday",
+                                  allow_nlp=["store_info"])
+    assert res['intent'] == 'get_store_hours'
+    assert 'sys_time' in {entity['type'] for entity in res['entities']}
+
+    res = kwik_e_mart_nlp.process("please can you tell me if springfield is "
+                                  "possibly open at this time on friday",
+                                  allow_nlp=["store_info"],
+                                  deny_nlp=["store_info.get_store_hours.sys_time"])
+    assert res['intent'] == 'get_store_hours'
+    assert 'sys_time' not in {entity['type'] for entity in res['entities']}
+
+    # If all entities in an intent are denied, that intent is also denied.
```

Comment:
Is that ideal though? Imposing this as a rule of thumb can lead to bad intent classification. If an application is not performing well on entities and a user tries to block entities out, this would cause the intent classifier to be erroneous as well.

Reply:
fixed

Reply:
In order for this to work, I had to break backwards compatibility on a little used API `allowed_nlp_classes`, which is an internal representation of the allowed NLP components in a dict structure. Here, if the entities are not explicitly added as sub-keys, we don't recognize them. However, the children of explicitly stated domain, intent and roles are recognized.